src/dhcpv6.c

```diff
@@ -0,0 +1,341 @@
+ *   disclaimer in the documentation and/or other materials provided
+ *   with the distribution.
+ *
+ *   Neither the name of the Cisco Systems, Inc. nor the names of its
+ *   contributors may be used to endorse or promote products derived
+ *   from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+/**
+ * \file dhcpv6.c
+ *
+ * \brief Dynamic Host Configuration Protocol (DHCP) for IPv6 awareness
+ *
+ */
+#include <stdlib.h>
+#include "dhcpv6.h"
+#include "p2f.h"
+#include "anon.h"
+#include "utils.h"
+#include "pkt.h"
+#include "err.h"
+
+#define DHCPV6_CLIENTID 1
+#define DHCPV6_SERVERID 2
+
+/* uncomment this to enable payload byte printing */
+//#define DHCPV6_DEBUG 1
+
+static const char *dhcpv6_msg_types[] = {
+    [1]  = "SOLICIT",
+    [2]  = "ADVERTISE",
+    [3]  = "REQUEST",
+    [4]  = "CONFIRM",
+    [5]  = "RENEW",
+    [6]  = "REBIND",
+    [7]  = "REPLY",
+    [8]  = "RELEASE",
+    [9]  = "DECLINE",
+    [10] = "RECONFIGURE",
+    [11] = "INFOREQUEST",
+    [12] = "RELAYFORWD",
+    [13] = "RELAYREPL"
+};
+
+static const char *dhcpv6_option_types[] = {
+    [1]  = "CLIENTID",
+    [2]  = "SERVERID",
+    [3]  = "IANA",
+    [4]  = "IATA",
+    [5]  = "IAADDR",
+    [6]  = "ORO",
+    [7]  = "PREFERENCE",
+    [8]  = "ELAPSEDTIME",
+    [11] = "AUTHENTICATION",
+    [12] = "UNICAST",
+    [13] = "STATUSCODE",
+    [14] = "RAPIDCOMMIT",
+    [15] = "USERCLASS",
+    [16] = "VENDORCLASS",
+    [17] = "VENDOROPTS",
+    [18] = "INTERFACEID",
+    [19] = "RECONFMSG",
+    [25] = "IAPD",
+    [26] = "IAPREFIX",
+    [32] = "INFOREFRESHTIME",
+    [82] = "SOLMAXRT",
+    [83] = "INFMAXRT"
+};
+
+static const char* dhcpv6_msg_to_string (uint16_t id) {
+    /*
+     * make sure the msg id is in the valid range
+     * from dhcpv6_msg_types definition above.
+     */
+    if ((id >= 1) && (id <= 13)) {
+        return dhcpv6_msg_types[id];
+    } else {
+        return NULL;
+    }
+}
+
+static const char* dhcpv6_option_to_string (uint16_t id) {
+    /*
+     * make sure the option id is in the valid range
+     * from dhcpv6_option_types definition above.
+     */
+    if (((id >= 1) && (id <= 19)) ||
```

Comment:
You may get garbage returned if id 9 and 10 are passed in.

Reply:
good catch, I will change